### PR TITLE
Dockersupport 1.1 workflow extension

### PR DESCRIPTION
**Pull request by lmanohara**
New workflow extensions for,
- Tenant Creation

### CODE REVIEW

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                "<email xmlns=\"http://wso2.org/bps/sample\">{6}</email>" +
+                "<active xmlns=\"http://wso2.org/bps/sample\">true</active>" +
+                "<successKey xmlns=\"http://wso2.org/bps/sample\">key</successKey>" +
+                "<createdDate xmlns=\"http://wso2.org/bps/sample\">2001-12-31T12:00:00</createdDate>" +
+                "<originatedService xmlns=\"http://wso2.org/bps/sample\">WSO2 Stratos Manager</originatedService>" +
+                "<usagePlan xmlns=\"http://wso2.org/bps/sample\">Demo</usagePlan>" +
+                "</p:CreateTenantRequest>";
+
+        String payLoadValue;
+        payLoadValue = MessageFormat.format(value, tenantCreationWorkflow.getTenantInfoBean().getAdmin(),
```
Comment (hisamith) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: No need to separate the declaration and the initialization.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Load configured class to the class loader and
+     * create instance of the particular class
+     */
+    public void loadWorkflowConfiguration() throws AppFactoryException {
+        if (log.isDebugEnabled()) {
+            String message = "Load configured class to class loader and create instance of that particular class "
+                    + "dynamically";
+            log.debug(message);
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: should check whether if log.isDebugEnabled

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                    + "object of the executor class for workflow type : "
+                    + WorkflowConstant.WorkflowType.TENANT_CREATION;
+            throw new AppFactoryException(message, e);
+        } catch (IllegalAccessException e) {
+            String message = "Unable to access the class" + executorClass + " for create new instance when accessing "
+                    + "the executor class for workflow type : " + WorkflowConstant.WorkflowType.TENANT_CREATION;
+            throw new AppFactoryException(message, e);
+        }
+    }
+
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: Why are we not logging errors?

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                            + WorkflowConstant.WorkflowType.TENANT_CREATION;
+            throw new AppFactoryException(message, e);
+        }
+
+        if (StringUtils.isBlank(executorClass)) {
+            executorClass = WorkflowConstant.TENANT_CREATION_DEFAULT_WORKFLOW_CLASS;
+            String message = "The tenant creation executor class is not define in appfactory.xml because tenant "
+                    + "creation default workflow is " + "executing. The tenant creation default workflow class is : "
+                    + executorClass;
+            log.warn(message);
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: why here is a warning log

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            String message = "Unable to find the bpel-policy.xml file";
+            throw new AppFactoryException(message);
+        } catch (XMLStreamException e) {
+            String message = "Unable to parse the xml";
+            throw new AppFactoryException(message, e);
+        } finally {
+            closeServiceClient(serviceClient);
+        }
+
+        log.info("The BPEL ran successfully to create tenant in the cloud. Tenant domain is : " +
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: this should be a debug log

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            throw new AppFactoryException(message, axisFault);
+        } catch (FileNotFoundException e) {
+            String message = "Unable to find the bpel-policy.xml file";
+            throw new AppFactoryException(message);
+        } catch (XMLStreamException e) {
+            String message = "Unable to parse the xml";
+            throw new AppFactoryException(message, e);
+        } finally {
+            closeServiceClient(serviceClient);
+        }
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: Should log the errors

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    "Can not initialize cloud manager for the tenant domain : " + workflowDTO.getTenantDomain();
+            throw new AppFactoryException(message, e);
+        }
+
+        log.info("The default tenant creation workflow executed successfully.Tenant domain is : " +
+                tenantCreationWorkflow.getTenantDomain() + ". Tenant Id is : " + tenantCreationWorkflow.getTenantId());
+
+    }
+
+}
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: Should log the errors and info log should be a debug log

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/AFTenantInitializer.java [hunk 8]
```diff
@@ -80,69 +74,72 @@ public static void initializeAFTenant(String tenantDomain) throws UserStoreExcep
...
+        tenantCreationWorkflowExecutor.execute(tenantCreationWorkflow);
+
     }
 
+    private static TenantInfoBean populateTenantInfoBean(String tenantDomain, int tenantId, Tenant tenant,
+            String firstName, String lastName) {
+        String successKey = "key";
+        String usagePlan = "Demo";
+        String originatedService = "WSO2 Stratos Manager";
+
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/AFTenantInitializer.java]: Remove hard coded strings

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            log.debug(message);
+        }
+
+        if (!(workflowDTO instanceof TenantCreationWorkflowDTO)) {
+            String message = "TenantCreationWorkflowDTO type is expected but unexpected type is passed to the execute"
+                    + " method for the tenant domain : " + workflowDTO.getTenantDomain();
+            if (log.isDebugEnabled()) {
+                log.debug(message);
+            }
+
```
Comment (hisamith) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: This should be an error log not a debug log

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java (hunk 4 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: Warning is explained in the message.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java (hunk 3 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: No need to log the error here. It is logged in https://github.com/wso2/product-af/pull/664/files#diff-032f69c60db0d971fd78c93eeb08da74R46

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java (hunk 4 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: Warn is explained in the message.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java (hunk 2 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/WorkflowExecutorFactory.java]: It is checked in above line.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java (hunk 6 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: When this method is called, errors are logging.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        return PolicyEngine.getPolicy(builder.getDocumentElement());
+    }
+
+    /**
+     * Execute tenant creation external workflowDTO process
+     *
+     * @param workflowDTO the tenant creation workflowDTO dto
+     * @throws AppFactoryException AppFactoryException exception throws if not success the execution
+     */
+    @Override public void execute(WorkflowDTO workflowDTO) throws AppFactoryException {
```
Comment (kasundsilva) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: @Override should come before method signature.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                serviceClient.cleanup();
+            }
+        } catch (AxisFault axisFault) {
+            String message = "Unable to clean up the resources which is used by the client";
+            log.error(message, axisFault);
+        }
+    }
+
+    private String createPayload(TenantCreationWorkflowDTO tenantCreationWorkflow) {
+        String value = "<p:CreateTenantRequest xmlns:p=\"http://wso2.org/bps/sample\">" +
```
Comment (danushkaf) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: Strings are immutable. Better using string builder if possible. By looking at it looks like a constant to me. WDYT?

Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationExternalWorkflowExecutor.java]: In here i am using MessageFormat.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
+/*
+ *   Copyright (c) 2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Kishanthan) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: Year should be 2016.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+import org.wso2.carbon.appfactory.core.services.AppFactoryTenantInfraStructureInitializerService;
+import org.wso2.carbon.appfactory.core.workflow.dto.TenantCreationWorkflowDTO;
+import org.wso2.carbon.appfactory.core.workflow.dto.WorkflowDTO;
+
+/**
+ * The executor class for the default workflow execution
+ */
+public class TenantCreationDefaultWorkflowExecutor implements WorkflowExecutor {
+
+    private static Log log = LogFactory.getLog(TenantCreationDefaultWorkflowExecutor.class);
```
Comment (Kishanthan) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: Best practice - logger instances should be final as we do not modify them after.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        } catch (AppFactoryException e) {
+            String message =
+                    "Can not initialize repository manager for the tenant domain : " + workflowDTO.getTenantDomain();
+            throw new AppFactoryException(message, e);
+        }
+
+        try {
+            if (log.isDebugEnabled()) {
+                String message =
+                        "Executing the initializeBuildManager for tenant domain : " + workflowDTO.getTenantDomain();
```
Comment (Kishanthan) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: Also I don't see a need to assign this "msg" to a string variable and then use it within the log. You could directly call the log with the msg as its used only once.

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java (hunk 12 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: This is fixed in following commit
https://github.com/wso2/product-af/pull/664/commits/6ddbe7623eeb6fdc99f0e16cadec68bdb1ffe0ea

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java (hunk 13 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: This is fixed in following commit.
https://github.com/wso2/product-af/pull/664/commits/6ddbe7623eeb6fdc99f0e16cadec68bdb1ffe0ea

File: modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java (hunk 14 above)
Comment (lmanohara) [inline modules/components/org.wso2.carbon.appfactory.core/src/main/java/org/wso2/carbon/appfactory/core/workflow/TenantCreationDefaultWorkflowExecutor.java]: This is fixed in following commit.
https://github.com/wso2/product-af/pull/664/commits/6ddbe7623eeb6fdc99f0e16cadec68bdb1ffe0ea

**Comment by manjulaRathnayaka** [comment]
@lmanohara 
We are closing this pull request because the project is no longer active and we will move this project to Attic soon. We are really thankful for your contributions.

**Pull request closed without merging**